File: bindings/python/hako_binary/offset_parser.py
#!/usr/bin/python
# -*- coding: utf-8 -*-


from select import select


def parse_offset(filepath):
    with open(filepath) as f:
        lines = f.readlines()
        return lines

def select_by_name(lines, name):
    for line in lines:
        if member_name(line) == name:
            return line
    return None

def member_name(data):
    return data.split(":")[2]

def member_type(data):
    return data.split(":")[3]

def is_primitive(data):
    return data.split(":")[1] == "primitive"

def is_single(data):
    return data.split(":")[0] == "single"

def is_array(data):
    return data.split(":")[0] == "array"

def is_varray(data):
    return data.split(":")[0] == "varray"

def array_size(data):
    if (is_single(data)):
        return 0
    if (is_varray(data)):
        return -1
    return int(data.split(":")[6])

def member_off(data):
    return int(data.split(":")[4])

def member_size(data):
    return int(data.split(":")[5])

File: bindings/elixir_project/lib/hako_binary/offset_parser.ex
defmodule OffsetParser do
  def parse_offset(filepath) do
    filepath
    |> File.read!()
    |> String.split("\n", trim: true)
  end

  def select_by_name(lines, name) do
    Enum.find(lines, fn line -> member_name(line) == name end)
  end

  def member_name(data) do
    data
    |> String.split(":")
    |> Enum.at(2)
  end

  def member_type(data) do
    data
    |> String.split(":")
    |> Enum.at(3)
  end

  def is_primitive(data) do
    data
    |> String.split(":")
    |> Enum.at(1)
    |> String.trim() == "primitive"
  end

  def is_single(data) do
    data
    |> String.split(":")
    |> Enum.at(0)
    |> String.trim() == "single"
  end

  def is_array(data) do
    data
    |> String.split(":")
    |> Enum.at(0)
    |> String.trim() == "array"
  end

  def is_varray(data) do
    data
    |> String.split(":")
    |> Enum.at(0)
    |> String.trim() == "varray"
  end

  def array_size(data) do
    cond do
      is_single(data) -> 0
      is_varray(data) -> -1
      true ->
        data
        |> String.split(":")
        |> Enum.at(6)
        |> String.to_integer()
    end
  end

  def member_off(data) do
    data
    |> String.split(":")
    |> Enum.at(4)
    |> String.to_integer()
  end

  def member_size(data) do
    data
    |> String.split(":")
    |> Enum.at(5)
    |> String.to_integer()
  end
end
